File: android/src/main/java/com/reactnativecouchbaselite/RNTask.kt
package com.reactnativecouchbaselite

import arrow.core.Either
import arrow.core.flatMap
import com.couchbase.lite.RNQuery
import com.couchbase.lite.SafeReadableMap
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap

data class RNTask(val group: String, val payload: ReadableMap) {
  companion object {
    fun run(obj: SafeReadableMap, eventEmitter: EventEmitter): Either<String, WritableMap> = obj
      .getString("group")
      .flatMap { group ->
        fun getPayload() = obj.getMap("payload")
        when (group) {
          "Replicator" -> getPayload().flatMap {
            RNReplicator.run(it, eventEmitter)
          }
          "Query" -> getPayload().flatMap {
            RNQuery.run(it)
          }
          else -> Either.left("unknown task group $group")
        }
      }
  }
}

File: ios/RNTask.swift
import Bow

class RNTask {
  static func run(_ obj: SafeDictionary) -> Either<String, [AnyHashable: Any]> {
    return obj.getString("group")
    .flatMap({ group in
      func getPayload() -> Either<String, SafeDictionary> {
        return obj.getDict("payload")
      }
      switch group {
      case "Replicator":
        return getPayload().flatMap({ RNReplicator.run($0) })
      default:
        return Either.left("unknown task group \(group)")
      }
    })^
  }
}
